File: src/sdk/contracts.ts
import { getSchema, listSchemas, SCHEMA_CONTRACT_VERSION, type JsonSchemaDefinition, type SchemaListEntry } from "../core/schemas.js";
import type { AuditResult } from "../commands/audit.js";
import type { AskSummary } from "../core/ask.js";
import type { BootstrapApplyResult, BootstrapContext, BootstrapValidationResult } from "../core/bootstrap.js";
import type { BootstrapProposeResult } from "../core/bootstrapPropose.js";
import type { ChangedFilesResult } from "../core/changed.js";
import type { CiReport } from "../core/ci.js";
import type { DoctorReport } from "../core/doctor.js";
import type { DriftResult } from "../core/drift.js";
import type { JsonCommandError } from "../core/errors.js";
import type { GraphQueryReport } from "../core/graphQuery.js";
import type { HandoffPacket } from "../core/handoff.js";
import type { InstallDoctorReport } from "../core/installDoctor.js";
import type { IntentListResult, IntentValidateResult } from "../commands/intent.js";
import type { IntentCheckResult, IntentVerifyResult } from "../core/intents.js";
import type { PrAssistantReport } from "../core/prAssistant.js";
import type { ReleaseDoctorReport } from "../core/releaseDoctor.js";
import type { DriftRepairReport } from "../core/repair.js";
import type { SchemaCoverageReport } from "../core/schemaCoverage.js";
import type { SemanticBuildResult, SemanticSearchResultPacket, SemanticStatusResult } from "../commands/semantic.js";
import type { SemanticEvaluationReport } from "../core/semanticIndex.js";
import type { LinkSuggestionEvalReport } from "../core/taskLinkEval.js";
import type { InstallCiWorkflowResult } from "../core/ciWorkflow.js";
import type { LastMutationSummary, MutationHistoryReport } from "../core/mutations.js";
import type { TaskSearchResult, SearchMode } from "../core/searchTasks.js";
import type { WorkPacket } from "../core/workPacket.js";
import type { JumpIndex, JumpPlan, JumpPlanStep, JumpTask, JumpTaskStatus, JumpVerificationRecord } from "../core/types.js";
import type { ExecutionState } from "../core/execution.js";

export const JUMPSPACE_CONTRACT_VERSION = SCHEMA_CONTRACT_VERSION;

export const JUMPSPACE_SCHEMA_NAMES = [
  "error",
  "schema.list",
  "schema.show",
  "schema.coverage",
  "task.list",
  "task.find",
  "task.find.compact",
  "task.audit",
  "task.last",
  "task.history",
  "task.handoff",
  "init.ci",
  "task.doctor",
  "release.doctor",
  "release.install-doctor",
  "task.context",
  "task.related",
  "task.related.compact",
  "task.plan.review",
  "task.plan.save",
  "task.plan.show",
  "task.plan.validate",
  "task.ready",
  "task.execute",
  "task.next",
  "task.step.complete",
  "task.status",
  "task.verify",
  "task.work",
  "task.ask",
  "task.ask.compact",
  "task.semantic.build",
  "task.semantic.status",
  "task.semantic.search",
  "task.semantic.eval",
  "task.query",
  "task.drift",
  "task.ci",
  "task.pr.comment",
  "intent.list",
  "intent.check",
  "intent.validate",
  "intent.verify",
  "task.repair",
  "task.link",
  "task.link.suggest",
  "task.link.eval",
  "task.bootstrap.context",
  "task.bootstrap.discover",
  "task.bootstrap.propose",
  "task.bootstrap.validate",
  "task.bootstrap.apply",
] as const;

export type JumpspaceSchemaName = (typeof JUMPSPACE_SCHEMA_NAMES)[number];

export type JumpspaceError = JsonCommandError;

export type JumpspaceErrorEnvelope = {
  ok: false;
  errors: JumpspaceError[];
};

export type JumpspaceCommandResult<T> = T | JumpspaceErrorEnvelope;

export type SchemaListCommandResult = {
  ok: true;
  contract_version: typeof JUMPSPACE_CONTRACT_VERSION;
  schemas: SchemaListEntry[];
};

export type SchemaShowCommandResult = {
  ok: true;
  contract_version: typeof JUMPSPACE_CONTRACT_VERSION;
  schema: JsonSchemaDefinition;
};
export type SchemaCoverageCommandResult = SchemaCoverageReport;

export type ListCommandResult = {
  tasks: JumpTask[];
};

export type FindCommandResult = {
  query: string;
  mode: SearchMode;
  results: TaskSearchResult[];
};

export type ContextCommandResult = {
  task: JumpTask;
  plan: JumpPlan | null;
  execution: ExecutionState | undefined;
};

export type PlanReviewCommandResult = ExecutionState;
export type PlanSaveCommandResult = {
  ok: true;
  task_id: string;
  plan: JumpPlan;
};
export type PlanShowCommandResult = {
  task_id: string;
  plan: JumpPlan;
};
export type PlanValidateCommandResult = {
  task_id: string;
  ok: boolean;
  issues: unknown[];
  errors: JumpspaceError[];
};
export type ReadyCommandResult = {
  tasks: ExecutionState[];
};
export type NextCommandResult = {
  task_id: string;
  steps: JumpPlanStep[];
};
export type ExecuteCommandResult = ExecutionState & {
  ok: boolean;
  forced: boolean;
  errors?: JumpspaceError[];
};
export type StepCompleteCommandResult = {
  ok: true;
  task_id: string;
  step: JumpPlanStep;
  plan: JumpPlan;
};
export type StatusCommandResult = {
  ok: true;
  task_id: string;
  status: Exclude<JumpTaskStatus, "verified">;
};
export type VerifyCommandResult = {
  ok: true;
  task_id: string;
  status: "verified";
  record: JumpVerificationRecord;
};

export type ChangedCommandResult = ChangedFilesResult;
export type DriftCommandResult = DriftResult;
export type RepairCommandResult = DriftRepairReport;
export type AuditCommandResult = AuditResult & {
  errors: AuditResult["issues"];
  warnings: AuditResult["issues"];
};
export type LastCommandResult = {
  ok: true;
  summary: LastMutationSummary;
};
export type HistoryCommandResult = MutationHistoryReport & {
  ok: true;
};
export type HandoffCommandResult = HandoffPacket;
export type DoctorCommandResult = DoctorReport;
export type ReleaseDoctorCommandResult = ReleaseDoctorReport;
export type ReleaseInstallDoctorCommandResult = InstallDoctorReport;
export type InitCiCommandResult = InstallCiWorkflowResult;
export type AskCommandResult = AskSummary;
export type SemanticBuildCommandResult = SemanticBuildResult;
export type SemanticStatusCommandResult = SemanticStatusResult;
export type SemanticSearchCommandResult = SemanticSearchResultPacket;
export type SemanticEvalCommandResult = SemanticEvaluationReport;
export type LinkEvalCommandResult = LinkSuggestionEvalReport;
export type QueryCommandResult = GraphQueryReport;
export type WorkCommandResult = WorkPacket;
export type CiCommandResult = CiReport;
export type PrCommentCommandResult = PrAssistantReport;
export type IntentListCommandResult = IntentListResult;
export type IntentCheckCommandResult = IntentCheckResult;
export type IntentValidateCommandResult = IntentValidateResult;
export type IntentVerifyCommandResult = IntentVerifyResult;
export type BootstrapContextCommandResult = BootstrapContext;
export type BootstrapProposeCommandResult = BootstrapProposeResult & {
  proposal_file?: string;
};
export type BootstrapValidateCommandResult = BootstrapValidationResult;
export type BootstrapApplyCommandResult = BootstrapApplyResult;
export type IndexJson = JumpIndex;

export function listSdkSchemas(): SchemaListEntry[] {
  return listSchemas();
}

export function getSdkSchema(name: JumpspaceSchemaName): JsonSchemaDefinition {
  const schema = getSchema(name);
  if (!schema) {
    throw new Error(`Unknown Jumpspace schema "${name}".`);
  }
  return schema;
}

export function isJumpspaceSchemaName(value: string): value is JumpspaceSchemaName {
  return (JUMPSPACE_SCHEMA_NAMES as readonly string[]).includes(value);
}

export function isJumpspaceErrorEnvelope(value: unknown): value is JumpspaceErrorEnvelope {
  if (!value || typeof value !== "object") {
    return false;
  }
  const candidate = value as { ok?: unknown; errors?: unknown };
  return candidate.ok === false && Array.isArray(candidate.errors);
}

export function assertOk<T>(result: JumpspaceCommandResult<T>): T {
  if (isJumpspaceErrorEnvelope(result)) {
    const message = result.errors.map((error) => error.message).join("; ") || "Unknown Jumpspace error.";
    throw new Error(message);
  }
  return result;
}

File: sdk/python/jumpspace_sdk/contracts.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Mapping, Sequence, TypeVar

CONTRACT_VERSION = 1

SCHEMA_NAMES: tuple[str, ...] = (
    "error",
    "schema.list",
    "schema.show",
    "schema.coverage",
    "task.list",
    "task.find",
    "task.find.compact",
    "task.audit",
    "task.last",
    "task.history",
    "task.handoff",
    "init.ci",
    "task.doctor",
    "release.doctor",
    "release.install-doctor",
    "task.context",
    "task.related",
    "task.related.compact",
    "task.plan.review",
    "task.plan.save",
    "task.plan.show",
    "task.plan.validate",
    "task.ready",
    "task.execute",
    "task.next",
    "task.step.complete",
    "task.status",
    "task.verify",
    "task.work",
    "task.ask",
    "task.ask.compact",
    "task.semantic.build",
    "task.semantic.status",
    "task.semantic.search",
    "task.semantic.eval",
    "task.query",
    "task.drift",
    "task.ci",
    "task.pr.comment",
    "intent.list",
    "intent.check",
    "intent.validate",
    "intent.verify",
    "task.repair",
    "task.link",
    "task.link.suggest",
    "task.link.eval",
    "task.bootstrap.context",
    "task.bootstrap.discover",
    "task.bootstrap.propose",
    "task.bootstrap.validate",
    "task.bootstrap.apply",
)

T = TypeVar("T")


@dataclass(frozen=True)
class CommandError:
    code: str
    message: str
    taskId: str | None = None
    path: str | None = None
    stepId: str | None = None


@dataclass(frozen=True)
class ErrorEnvelope:
    errors: Sequence[CommandError]
    ok: bool = False


@dataclass(frozen=True)
class SchemaListEntry:
    name: str
    command: str
    description: str


@dataclass(frozen=True)
class SchemaDefinition:
    name: str
    command: str
    description: str
    schema: Mapping[str, Any]


@dataclass(frozen=True)
class SchemaListResult:
    schemas: Sequence[SchemaListEntry]
    ok: bool = True
    contract_version: int = CONTRACT_VERSION


@dataclass(frozen=True)
class SchemaShowResult:
    schema: SchemaDefinition
    ok: bool = True
    contract_version: int = CONTRACT_VERSION


@dataclass(frozen=True)
class JumpTask:
    id: str
    title: str
    type: str
    status: str
    doc: Mapping[str, Any]
    spec: str
    code: Sequence[str] = field(default_factory=tuple)
    tests: Sequence[str] = field(default_factory=tuple)
    depends_on: Sequence[str] = field(default_factory=tuple)
    module: str | None = None
    space: str | None = None
    keywords: Sequence[str] = field(default_factory=tuple)
    gaps: Sequence[str] = field(default_factory=tuple)


@dataclass(frozen=True)
class ListResult:
    tasks: Sequence[JumpTask | Mapping[str, Any]]


@dataclass(frozen=True)
class FindResult:
    query: str
    mode: str
    results: Sequence[Mapping[str, Any]]


@dataclass(frozen=True)
class ContextResult:
    task: JumpTask | Mapping[str, Any]
    plan: Mapping[str, Any] | None
    execution: Mapping[str, Any] | None


@dataclass(frozen=True)
class WorkResult:
    task: JumpTask | Mapping[str, Any]
    plan: Mapping[str, Any]
    next_steps: Sequence[Mapping[str, Any]]
    ok: bool = True
    packet_version: int = 1
    schemas: Mapping[str, str] = field(default_factory=dict)


@dataclass(frozen=True)
class ChangedResult:
    since: str
    files: Sequence[Mapping[str, Any]]
    ok: bool = True


@dataclass(frozen=True)
class GraphQueryResult:
    query: Mapping[str, Any]
    results: Sequence[Mapping[str, Any]]
    unanswered_constraints: Sequence[Mapping[str, Any]]
    ok: bool = True


@dataclass(frozen=True)
class HistoryResult:
    history_path: str
    total: int
    returned: int
    filters: Mapping[str, Any]
    entries: Sequence[Mapping[str, Any]]
    ok: bool = True


@dataclass(frozen=True)
class HandoffResult:
    packet_version: int
    generated_at: str
    status: str
    filters: Mapping[str, Any]
    summary: Mapping[str, Any]
    recent_mutations: Mapping[str, Any]
    last_mutation: Mapping[str, Any] | None
    touched_files: Sequence[str]
    task_ids: Sequence[str]
    config_changes: Sequence[str]
    mutation_warnings: Sequence[Mapping[str, Any]]
    health: Mapping[str, Any]
    task: Mapping[str, Any] | None
    suggested_commands: Sequence[str]
    schemas: Mapping[str, str]
    ok: bool = True


@dataclass(frozen=True)
class SemanticBuildResult:
    index_path: str
    task_count: int
    document_count: int
    source_index: Mapping[str, Any]
    backend: Mapping[str, Any]
    config_updated: bool
    ok: bool = True


@dataclass(frozen=True)
class SemanticStatusResult:
    enabled: bool
    path: str
    exists: bool
    ready: bool
    stale: bool
    issues: Sequence[Mapping[str, Any]]
    source_index: Mapping[str, Any] | None
    backend: Mapping[str, Any] | None
    document_count: int
    ok: bool = True


@dataclass(frozen=True)
class SemanticSearchResult:
    query: str
    index_path: str
    backend: Mapping[str, Any]
    results: Sequence[Mapping[str, Any]]
    ok: bool = True


@dataclass(frozen=True)
class SemanticEvalResult:
    query_count: int
    summary: Mapping[str, Any]
    active_backend: Mapping[str, Any] | None
    results: Sequence[Mapping[str, Any]]
    ok: bool = True


@dataclass(frozen=True)
class LinkEvalResult:
    suite: str
    fixture_path: str | None
    case_count: int
    summary: Mapping[str, Any]
    cases: Sequence[Mapping[str, Any]]
    ok: bool = True


@dataclass(frozen=True)
class ReleaseDoctorResult:
    checked_at: str
    status: str
    summary: Mapping[str, Any]
    package: Mapping[str, Any]
    package_dry_run: Mapping[str, Any]
    schemas: Mapping[str, Any]
    registry: Mapping[str, Any]
    local_blockers: Sequence[Mapping[str, Any]]
    local_warnings: Sequence[Mapping[str, Any]]
    external_warnings: Sequence[Mapping[str, Any]]
    ok: bool = True


@dataclass(frozen=True)
class BootstrapProposeResult:
    proposal: Mapping[str, Any]
    validation: Mapping[str, Any]
    summary: Mapping[str, Any]
    ok: bool = True
    propose_version: int = 1
    mode: str = "deterministic_extraction"
    agent_generated: bool = False
    human_approval_required: bool = True
    proposal_file: str | None = None


@dataclass(frozen=True)
class BootstrapApplyResult:
    dry_run: bool
    applied: Sequence[Mapping[str, Any]]
    config_paths_added: Sequence[str]
    ok: bool = True


def is_schema_name(value: str) -> bool:
    return value in SCHEMA_NAMES


def is_error_envelope(value: Mapping[str, Any]) -> bool:
    return value.get("ok") is False and isinstance(value.get("errors"), list)


def assert_ok(result: T | Mapping[str, Any]) -> T | Mapping[str, Any]:
    if isinstance(result, Mapping) and is_error_envelope(result):
        messages = [str(error.get("message", "")) for error in result.get("errors", []) if isinstance(error, Mapping)]
        raise ValueError("; ".join(message for message in messages if message) or "Unknown Jumpspace error.")
    return result
